Docs : Add "Python Scripting API" section and "The Python Editor" article

medubelko opened the pull request:
This is coming on the tail of expanding the existing scripting related articles and moving everything into its own section. Once merged, all the scripting tutorials can be moved into this section.

@themissingcow See the docs I shared in Drive.

### Checklist ###

- [x] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/master/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable.
- [x] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [x] My code follows the Gaffer project's prevailing coding style and conventions.

themissingcow commented:
Does this need a `.gitignore` update?

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+
+# The Python Editor #
+
+The _Python Editor_ is a type of editor in the main Gaffer application that lets you build and modify the node graph, test expressions, and inspect data by inputting and executing Python code. In the default layout, it resides in the bottom-right panel, in a tab next to the _Hierarchy View_.
```
If the audience potentially includes developers, it might be good to be a little more precise at the beginning here, how about:

_The _Python Editor_ is a type of editor in the main Gaffer application that lets you execute arbitrary python code within the UI's interpreter. This allows you to (amongst other things) build and modify the node graph and inspect data...._

> test expressions

Be careful here, the expression interpreter is not the same as the Python Editor's interpreter. As such, some things you can do in the Editor will not work in an expression and visa versa. For example, the expression engine defines additional `parent` and `context` variables in the global scope which have no equivalent in the Editor. If anything, its worth including a _warning_ here that there are differences between the two (and differences beyond those I mentioned here)!

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+## Instructions ##
+
+
+### Executing Code ###
```
Sentence case?

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+> When referring to objects or scene locations, remember to consider the selected node's context. Locations in the scene hierarchy can vary from node to node, depending on what happens to the hierarchy through the node graph. If you refer to a location from one node, but use a different node's context in your query, the reference may point to a location that doesn't exist.
+
+
+## See Also ##
```
Sentence case?

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/index.md:
```diff
@@ -0,0 +1,40 @@
+Do not confuse scripts with active node graphs or expressions: when a node graph is open in Gaffer, or in the middle of executing, its nodes and plugs are stored in memory as C++ objects. All scripts do is instruct Gaffer on how to **build** graphs. Serialization is kept deliberately simple to limit complexity, and is why a node graph cannot dynamically add or remove nodes and plugs to itself dur ...
+
+
+## Python Scripting in Graphs ##
```
lowercase 'S'?

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+
+The _Python Editor_ is split into two areas. The bottom-half is the input field. The top-half is the output. The input field of the _Python Editor_ functions like a basic plain text editor. You can type, select, cut, copy, paste, and execute code. 
+
+![A Python Editor with its input and the output field](images/interfacePythonEditorOutput.png "A Python Editor with its input and the output field")
```
It might just be my local build, but the caption seems to be inline with the image:
![image](https://user-images.githubusercontent.com/896779/58035806-7941bf80-7b21-11e9-96a9-7af936e5bbd9.png)

medubelko replied:
Which version of ReadTheDocs/Recommonmark did you build with? This looks like [a certain bug](https://github.com/rtfd/recommonmark/issues/88) in more recent versions of Recommonmark.

themissingcow replied:
Looks like:
```
recommonmark                           0.5.0   
Sphinx                                 1.8.5   
sphinx-rtd-theme                       0.4.3 
```  
Ill try the container build and see how it looks. Thanks!

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+
+1. Enter code into the input field.
+2. Highlight any portion of the input you wish to preserve for reuse.
+3. Hit <kbd>Control</kbd> + <kbd>Enter</kbd>.
```
> 2. Highlight any portion of the input you wish to preserve for reuse.

The behaviour is actually a little different to this:
 - If nothing is highlighted the entire input is evaluated then cleared.
 - If a any of the input is highlighted _only the highlighted code is evaluated_ and the entire input is preserved.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+
+### Grabbing data from other editors ###
+
+The _Python Editor_ can refer to the same nodes, plugs, objects, scene locations, and inspected scene data accessible from the various other editors. However, this requires inputting precise dictionary syntax, types, or code. Instead of referring to these kinds of data manually, you can use a shortcut: simply drag the node, plug, or location from its source editor into the _Python Editor's_ input ...
```
> The _Python Editor_ can refer to the same nodes...

I'm not sure 'referring' is the right word in this context. From a programming perspective, referring is effectively a very specific type of operation (ie, pointing to something thats somewhere else). This may not be what you're doing (see below). How about something like?

_You may wish to access the same nodes and plugs as other editors or create data similar to that in your scene..._   

> of referring to these kinds of data manually

As above, in may cases you're not 'referring' to anything, you're actually creating new objects, for example, dragging an Attribute's value from the Scene Inspector will actually create new data, entirely unrelated to the source when you evaluate your expression. Dragging location(s) from the Hierarchy View simply creates a list of strings of paths to the selected locations, not the locations themselves (they don't really 'exist' per-se to be referenced to).

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+Node(s)                  *Graph Editor*    Middle-click and drag  |nodes|
+Plug                     *Node Editor*     Click and drag         |plug|
+Plug value               *Node Editor*     Middle-click and drag  |values|
+Object(s)                *Viewer*          Click and drag         |objects|
```
> Object(s)

This is just a list of strings of paths to the selected locations. see below. There isn't an 'Object' data type.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,72 @@
+Plug                     *Node Editor*     Click and drag         |plug|
+Plug value               *Node Editor*     Middle-click and drag  |values|
+Object(s)                *Viewer*          Click and drag         |objects|
+Scene location(s)        *Hierarchy View*  Click and drag         |objects|
```
> Scene location(s)

There isn't a Scene location data type, instead you get a list of strings of paths to the selected locations. These can be used for various purposes, but they're not anything fancy in terms of a data type.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
````diff
@@ -0,0 +1,72 @@
+
+```eval_rst
+======================== ================= ====================== ============
+Data Type                Source Editor     Shortcut               Cursor
````
> Data Type

To be accurate, the below aren't (all) data types, they're 'logical things'.   In reality, these are: 

| Old |  Actual type |
|-|-|
| Node | `Gaffer.Node` or `[Gaffer.Node]` |
| Plug | `Gaffer.Plug` |
| Plug value | `float`/`string`/`int`,`bool`/`IECore.Data`, etc... |
| Object(s) | `[str]` |
| Scene locations(s) | `[str]` |
| Scene data | `float`/`string`/`int`,`bool`/`IECore.Data`, etc... |

I wonder if its worth re-arranging so that it lists types you might want against where you get them from? if you don't want to get into the actual types, then maybe change the heading? But I think there is definitely value in knowing that if I drag an object from the viewer I get a list of paths.

medubelko commented on doc/source/WorkingWithThePythonScriptingAPI/index.md:
```diff
@@ -0,0 +1,40 @@
+
+Gaffer node graphs are stored on disk as Python scripts with the `.gfr` file extension. References stored on disk are also Python scripts, but with the `.grf` extension.
+
+Do not confuse scripts with active node graphs or expressions: when a node graph is open in Gaffer, or in the middle of executing, its nodes and plugs are stored in memory as C++ objects. All scripts do is instruct Gaffer on how to **build** graphs. Serialization is kept deliberately simple to limit complexity, and is why a node graph cannot dynamically add or remove nodes and plugs to itself dur ...
```
I've removed it, but FYI I wrote the last sentence in reaction to a question on the gaffer-dev list about graphs that self-restructure during dispatch.

themissingcow replied:
Cool, thanks. How about making it something we talk about first-class wherever we describe the node graph and its execution in general?

medubelko commented:
Hi @themissingcow, thanks for the input. I admit my doc for this dates back to before I understood things better, hence some of the more glaring issues. Rather than spam everyone with emails, I've put all my proposed solutions in the Google docs I shared with you. We can continue the review there.

medubelko commented:
@themissingcow Updated and rebased. Let me know if there are any points I missed. I added a word or two here and there, and a .gif, but the main changes are the edits we worked on.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/index.md:
```diff
@@ -0,0 +1,30 @@
+With its core architecture built from a combination of C++ and Python, Gaffer ships with a convenient Python scripting API. A significant portion of the project has hooks in Python: nodes are exposed to Python in the main application, the node graph file formats are Python scripts, much of the GUI is written in Python, and the various Gaffer applications are configured by Python scripts. Through  ...
+
+> Important :
+> With regards to node graphs, do not confuse the use of the scripting API with expressions inside Expression nodes. The scripting API can be applied in the main application to interactively modify, inspect, test, and debug a node graph. Expressions, on the other hand, are bits of code stored on Expression nodes **inside** a graph, that manipulate data as part of the graph's function.
```
> Expression nodes **inside** a graph, that manipulate data as part of the graph's function

Expressions don't directly manipulate data, they're connected to Plugs to determine their value during compute. How about: 

"... that allow you to dynamically change a plugs value."

... or something to that effect?

medubelko replied:
Sure, let's just make it a little clearer:

> ... that allow you to dynamically change plug values during the graph's execution.

On this topic: When using the term "data" to relate "information that flows through the graph," do we want a hard separation between data that results from querying `out` plugs with context, and "actual" 2D/3D data, as it were? Do we want to separate them for developers?

themissingcow replied:
There is no difference between the two! The 'actual' data _is_ just data from plugs queried with a context (hence some of my previous comments about not talking about 'data' in certain places).

themissingcow replied:
The re-world looks good though, thanks!

themissingcow replied:
Just thinking about the above, maybe the confusion is that internally, Expressions use python/OSL to derive the value of _their_ out plugs, these plugs are then connected to the 'target' plugs on some other Node. Do during compute when an affected node queries the value of its plug to determine what it should do, it will result in the value being pulled from the connected expression output - just as any other plug-plug connection. If that makes sense?

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,78 @@
+
+![Errors in a Python Editor's output field](images/interfacePythonEditorError.png "Errors in a Python Editor's output field")
+
+> Tip :
```
Would this tip be better placed later in the article? We haven't demonstrated any commands yet and it talks about adding nodes to the graph. It feels like it should come after 'Executing code' at least?

medubelko replied:
At this stage, we're just introducing the editor's elements. "Seeing what an error looks like" isn't a task. Admittedly this might be better served in a general UI article (of which we don't have any yet), but I'd like to leave it.

themissingcow replied:
Sorry this is GitHubs's irritating context quoting. My comment was really on the tip about having to import modules to be able to add nodes to a grab. Error docs = good!

medubelko replied:
Ah, that makes more sense. Good point.

medubelko replied:
I ended up moving this part into the main text, and explaining how each _Python Editor_ is a separate instance.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/ThePythonEditor/index.md:
```diff
@@ -0,0 +1,78 @@
+> Tip :
+> A useful debugging trick is to highlight any part of the input code in order to preserve the whole input for reuse.
+
+
```
It might be nice to add some links to the Scripting Tutorials/Reference here?

medubelko replied:
The plan is that as soon as this is merged, I'll make a separate PR that moves all the scripting tutorials into this section, and then add links. Doing it this way just saves the headache of doubling up.

themissingcow commented on doc/source/WorkingWithThePythonScriptingAPI/index.md:
```diff
@@ -0,0 +1,30 @@
+## Python scripting API basics ##
+
+- [The Python Editor](ThePythonEditor/index.md)
+
```
Tutorial links here too?

medubelko replied:
See above.

medubelko commented:
@themissingcow Added some commits to address your feedback. Let me know if it's ready to rebase + merge.

themissingcow commented:
LGTM, thanks!
